packages/actor-http-retry/lib/ActorHttpRetry.ts

```diff
@@ -0,0 +1,178 @@
+    }
+    const retryCount = action.context.get<number>(KeysHttp.httpRetryCount);
+    if (!retryCount || retryCount < 1) {
+      return failTest(`${this.name} requires a retry count greater than zero to function`);
+    }
+    return passTest({ time: 0 });
+  }
+
+  public async run(action: IActionHttp): Promise<IActorHttpOutput> {
+    const url = ActorHttp.getInputUrl(action.input);
+
+    // Attempt once + the number of retries specified by the user
+    const attemptLimit = action.context.getSafe<number>(KeysHttp.httpRetryCount) + 1;
+    const retryDelay = action.context.get<number>(KeysHttp.httpRetryDelay) ?? 0;
+    const retryStatusCodes = action.context.get<number[]>(KeysHttp.httpRetryStatusCodes);
+
+    // This is declared outside the loop so it can be used for the final error message
+    for (let attempt = 1; attempt <= attemptLimit; attempt++) {
+      if (url.host in this.activeDelays) {
+        this.logDebug(action.context, 'Delaying request due to host rate limit', () => ({
+          url: url.href,
+          delayedUntil: this.activeDelays[url.host].date.toISOString(),
+          currentAttempt: `${attempt} / ${attemptLimit}`,
+        }));
+        await ActorHttpRetry.waitUntil(this.activeDelays[url.host].date);
+      } else if (attempt > 1) {
+        await ActorHttpRetry.waitUntil(new Date(Date.now() + retryDelay));
+      }
+
+      const response = await this.mediatorHttp.mediate({
+        ...action,
+        context: action.context.set(ActorHttpRetry.keyWrapped, true),
+      });
+
+      if (response.ok) {
+        return response;
+      }
+
+      if (retryStatusCodes && retryStatusCodes.includes(response.status)) {
+        this.logDebug(action.context, 'Status code in force retry list, forcing retry', () => ({
+          url: url.href,
+          status: response.status,
+          statusText: response.statusText,
+          currentAttempt: `${attempt} / ${attemptLimit}`,
+        }));
+        continue;
+      }
+
+      if (response.status === 504) {
+        // When the server is acting as a proxy and the source times it, it makes sense to retry
+        // with the hope that the source server replies within the timeout
+        this.logDebug(action.context, 'Received proxy timeout', () => ({
+          url: url.href,
+          status: response.status,
+          statusText: response.statusText,
+          currentAttempt: `${attempt} / ${attemptLimit}`,
+        }));
+        continue;
+      }
+
+      if (response.status === 429 || response.status === 503) {
+        // When the server reports a rate limit or temporary unavailability,
+        // then wait for the amount of time or for the specific date/time specified in Retry-After,
+        // or wait for the default user-specified time if this header is not provided by the server
+        const retryAfter = ActorHttpRetry.parseRetryAfterHeader(
+          response.headers.get('retry-after') ?? retryDelay.toString(10),
```

Comment:
We need 2 additional checks here:

- A sanity check to make sure the delay makes sense. As to not make the query engine wait for literal ages in the case of server misconfigurations. Maybe we need an upper limit of what we accept for this value.
- A try-catch around this parsing, which we emit as a warning. As to not make the engine crash if the server emits some illegal value.

Reply:
These have been incorporated now:
* `httpRetryDelayLimit` is the upper limit for the wait time
* the actor will ignore the server-provided header and log a warning when the value is bad